meshroom/ui/app.py

```diff
@@ -328,38 +335,89 @@ def reloadTemplateList(self):
+            for node in fileData:
+                if fileData[node]["nodeType"] == "CameraInit" and fileData[node]["inputs"].get("viewpoints"):
+                    if len(fileData[node]["inputs"]["viewpoints"]) > 0:
+                        thumbnail = fileData[node]["inputs"]["viewpoints"][0]["path"]
+                        break
+        except FileNotFoundError:
+            pass
+
+        return thumbnail
+
+    def _getRecentProjectFiles(self) -> list[dict[str, str]]:
+        """
+        Read the list of recent project files from the QSettings, retrieve their filepath, and if it exists, their
+        thumbnail.
+
+        Returns:
+            list[dict[str, str]]: The list containing dictionaries of the form {"path": "/path/to/project/file",
+                                                                                "thumbnail": "/path/to/thumbnail"}
+                                  based on the recent projects stored in the QSettings.
+        """
         projects = []
         settings = QSettings()
         settings.beginGroup("RecentFiles")
         size = settings.beginReadArray("Projects")
         for i in range(size):
             settings.setArrayIndex(i)
-            p = settings.value("filepath")
-            thumbnail = ""
-            if p:
-                # get the first image path from the project
-                try:
-                    with open(p) as file:
-                        file = json.load(file)
-                        # find the first camerainit node
-                        file = file["graph"]
-                        for node in file:
-                            if file[node]["nodeType"] == "CameraInit" and file[node]["inputs"].get("viewpoints"):
-                                if len(file[node]["inputs"]["viewpoints"]) > 0:
-                                    thumbnail = file[node]["inputs"]["viewpoints"][0]["path"]
-                                    break
-                except FileNotFoundError:
-                    pass
-                p = {"path": p, "thumbnail": thumbnail}
+            path = settings.value("filepath")
+            if path:
+                p = {"path": path, "thumbnail": self._retrieveThumbnailPath(path)}
                 projects.append(p)
         settings.endArray()
         settings.endGroup()
         return projects
 
+    @Slot()
+    def updateRecentProjectFilesThumbnails(self) -> None:
+        """
+        If there are thumbnails that may be retrievable (meaning the list of projects has been updated minimally),
+        update the list of recent project files by reading the QSettings and retrieving the thumbnails if they are
+        available.
+        """
+        if not self._updatedRecentProjectFilesThumbnails:
+            self._updateRecentProjectFilesThumbnails()
+            self._updatedRecentProjectFilesThumbnails = True
+
+    def _updateRecentProjectFilesThumbnails(self) -> None:
+        projects = self._recentProjectFiles
```

Comment:
I'm not sure to get why we have this `projects` variable (instead of working directly with the class member).

Reply:
Done